app/src/main/java/com/amaze/filemanager/database/UtilitiesDatabase.java

```diff
@@ -0,0 +1,352 @@
+              "INSERT INTO " + backupTable + " SELECT * FROM " + TABLE_HISTORY + " group by path;");
+          database.execSQL("DROP TABLE " + TABLE_HISTORY + ";");
+          database.execSQL("ALTER TABLE " + backupTable + " RENAME TO " + TABLE_HISTORY + ";");
+
+          backupTable = TEMP_TABLE_PREFIX + TABLE_HIDDEN;
+          database.execSQL(queryHidden.replace(TABLE_HIDDEN, backupTable));
+          database.execSQL(
+              "INSERT INTO " + backupTable + " SELECT * FROM " + TABLE_HIDDEN + " group by path;");
+          database.execSQL("DROP TABLE " + TABLE_HIDDEN + ";");
+          database.execSQL("ALTER TABLE " + backupTable + " RENAME TO " + TABLE_HIDDEN + ";");
+
+          backupTable = TEMP_TABLE_PREFIX + TABLE_LIST;
+          database.execSQL(queryList.replace(TABLE_LIST, backupTable));
+          database.execSQL(
+              "INSERT INTO " + backupTable + " SELECT * FROM " + TABLE_LIST + " group by path;");
+          database.execSQL("DROP TABLE " + TABLE_LIST + ";");
+          database.execSQL("ALTER TABLE " + backupTable + " RENAME TO " + TABLE_LIST + ";");
+
+          backupTable = TEMP_TABLE_PREFIX + TABLE_GRID;
+          database.execSQL(queryGrid.replace(TABLE_GRID, backupTable));
+          database.execSQL(
+              "INSERT INTO " + backupTable + " SELECT * FROM " + TABLE_GRID + " group by path;");
+          database.execSQL("DROP TABLE " + TABLE_GRID + ";");
+          database.execSQL("ALTER TABLE " + backupTable + " RENAME TO " + TABLE_GRID + ";");
+
+          backupTable = TEMP_TABLE_PREFIX + TABLE_BOOKMARKS;
+          database.execSQL(queryBookmarks.replace(TABLE_BOOKMARKS, backupTable));
+          database.execSQL(
+              "INSERT INTO "
+                  + backupTable
+                  + " SELECT * FROM "
+                  + TABLE_BOOKMARKS
+                  + " group by path;");
+          database.execSQL("DROP TABLE " + TABLE_BOOKMARKS + ";");
+          database.execSQL("ALTER TABLE " + backupTable + " RENAME TO " + TABLE_BOOKMARKS + ";");
+
+          backupTable = TEMP_TABLE_PREFIX + TABLE_SMB;
+          database.execSQL(querySmb.replace(TABLE_SMB, backupTable));
+          database.execSQL(
+              "INSERT INTO " + backupTable + " SELECT * FROM " + TABLE_SMB + " group by path;");
+          database.execSQL("DROP TABLE " + TABLE_SMB + ";");
+          database.execSQL("ALTER TABLE " + backupTable + " RENAME TO " + TABLE_SMB + ";");
+
+          backupTable = TEMP_TABLE_PREFIX + TABLE_SFTP;
+          database.execSQL(querySftp.replace(TABLE_SFTP, backupTable));
+          database.execSQL(
+              "INSERT INTO " + backupTable + " SELECT * FROM " + TABLE_SFTP + " group by path;");
+          database.execSQL("DROP TABLE " + TABLE_SFTP + ";");
+          database.execSQL("ALTER TABLE " + backupTable + " RENAME TO " + TABLE_SFTP + ";");
+        }
+      };
+
+  private static final Migration MIGRATION_3_4 =
+      new Migration(3, 4) {
```

Comment:
I too can't understand why... but in the early days of this PR it seems Room requires all tables to have a `_id` column; and add the `_id` column with primary key attribute didn't work on my devices (should be Oneplus 2 with AOSPExtended 7.1.2).

Reply:
No, what I mean is, you're essentially doing `MIGRATION_2_3` here again, the only difference is adding a new constraint on `_id` which is `NOT NULL`. Instead of creating new table and transferring data for that, you can simply add this constraint as follows:
`ALTER TABLE table_name MODIFY _id INTEGER NOT NULL;`
Please test if primary key constraint is retained after doing this.

Reply:
Crash on start up.

And finally my browser history reminded me the reason why I needed to go all the hussle:

https://stackoverflow.com/a/4007086

And [officially, SQL Features That SQLite Does Not Implement](https://www.sqlite.org/omitted.html):

Complete ALTER TABLE support | Only the RENAME TABLE, ADD COLUMN, and RENAME COLUMN variants of the    ALTER TABLE command are supported.  Other kinds of ALTER TABLE operations   such as   DROP COLUMN, ALTER COLUMN, ADD CONSTRAINT, and so forth are omitted.
-- | --

Reply:
Got it.. thanks for checking. 😶